File: 1605-minimum-number-of-days-to-make-m-bouquets/minimum-number-of-days-to-make-m-bouquets.py
class Solution:
    def minDays(self, bloomDay: List[int], m: int, k: int) -> int:
        lo, hi = 1, max(bloomDay)
        ans = -1
        def check(days):
            flwrs = 0
            bqts = 0
            for b in bloomDay:
                if b <= days:
                    flwrs += 1
                    if flwrs == k:
                        flwrs = 0
                        bqts += 1
                else:
                    flwrs = 0
            return bqts >= m
        while lo <= hi:
            mid = (lo + hi) // 2
            if check(mid):
                ans = mid
                hi = mid - 1
            else:
                lo = mid + 1
        return ans

File: 1605-minimum-number-of-days-to-make-m-bouquets/minimum-number-of-days-to-make-m-bouquets.cpp
class Solution {
public:
    bool    check(vector<int>& bloomDay, int m, int k, int day) {
        int flowers = 0, bouquets = 0;
        for (auto b : bloomDay) {
            if (b <= day) {
                flowers += 1;
                if (flowers == k) {
                    bouquets += 1;
                    flowers = 0;
                }
            }
            else {
                flowers = 0;
            }
        }
        return bouquets >= m;
    }
    int     minDays(vector<int>& bloomDay, int m, int k) {
        int res = -1;
        int low = *std::min_element(bloomDay.begin(), bloomDay.end());
        int high = *std::max_element(bloomDay.begin(), bloomDay.end());
        int mid; 
        while (low <= high) {
            mid = low + (high - low) / 2;
            if (check(bloomDay, m, k, mid)) {
                res = mid;
                high = mid - 1;
            }
            else {
                low = mid + 1;
            }
        }
        return res;
    }
};